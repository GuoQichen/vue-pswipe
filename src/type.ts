import Vue from 'vue'
import { Options, Item } from 'photoswipe/dist/photoswipe-ui-default'

/**
 * v-pswp directive config options
 */
export interface PswpDirectiveOptions {
    /**
     * path to image
     */
    src: string
    /**
     * image size, 'width x height', eg: '100x100'
     */
    size?: string
    /**
     * small image placeholder,
     * main (large) image loads on top of it,
     * if you skip this parameter - grey rectangle will be displayed,
     * try to define this property only when small image was loaded before
     */
    msrc?: string
    /**
     * used by Default PhotoSwipe UI
     * if you skip it, there won't be any caption
     */
    title?: string
    /**
     * to make URLs to a single image look like this: http://example.com/#&gid=1&pid=custom-first-id
     * instead of: http://example.com/#&gid=1&pid=1
     * enable options history:true, galleryPIDs:true and add pid (unique picture identifier)
     */
    pid?: string | number
}

export interface PswpItem extends Item {
    el: HTMLElement
    src: string
    msrc?: string
    pid?: number | string
    verticalRotated?: boolean
}

export interface ManualImgItem extends Partial<Item> {
    src: string
}

export type ManualHtmlItem = Partial<Item> 

export interface CurrentPswpItem extends PswpItem {
    container: HTMLElement
    loaded: boolean
}

export type PswpOptions = Options

export interface OpenPhotoSwipeArgs {
    index: number | string
    fromURL?: boolean
    thumbEls?: HTMLElement[]
}

export interface BeforeOpenEvent {
    index: number
    target: HTMLElement
    items: PswpItem[]
    options: PswpOptions
}

export interface Size {
    w: number
    h: number
}

export interface ManualCreateArgs {
    items: ManualOpenItem[]
    options?: PswpOptions
}

interface CreatePhotoSwipeArgs extends ManualCreateArgs {
    context?: Vue
}

export interface PswpProps {
    options: PswpOptions
    auto: boolean
    bubble: boolean
    lazy: boolean
    filter: Function
    rotate: boolean
}

// types
export type Fn = (...args: any[]) => void

export type Pswp = PhotoSwipe<PswpOptions>

export type BeforeOpen = (continued?: boolean) => void

export type Filter = (img: HTMLImageElement) => boolean

export type FindIndex = <T>(
    array: T[],
    predicate: (item: T, idx: number) => boolean
) => number

export type Closest = (
    el: Node | null,
    predicate: (el: HTMLElement) => boolean
) => HTMLElement | false

export type Get = <T>(
    context: Record<string, any>,
    path: string,
    defaultValue: T
) => T

export type Single = <T>(fn: Function) => (...args: any[]) => T

export type BindEvent = (pswp: Pswp, context?: Vue) => void

export type CreatePhotoSwipe = (arg: CreatePhotoSwipeArgs) => Pswp

export type HandleWithoutSize = (pswp: Pswp) => void

export type GetContainSize = (
    areaWidth: number,
    areaHeight: number,
    width: number,
    height: number
) => Size

export type RotateDirection = 'left' | 'right'

export type StyleKey = Exclude<keyof CSSStyleDeclaration, number | 'length' | 'parentRule'>

export type ManualOpenItem = ManualImgItem | ManualHtmlItem
