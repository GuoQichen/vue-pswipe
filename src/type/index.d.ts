import PhotoSwipeUI_Default from 'photoswipe/dist/photoswipe-ui-default'

export interface Options extends PhotoSwipeUI_Default.Options {
    galleryPIDs?: number
}

export type Filter = (img: HTMLImageElement) => boolean
