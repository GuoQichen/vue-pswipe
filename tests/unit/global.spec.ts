/* eslint-disable import/first */
import {
    PhotoSwipe,
    PhotoSwipeMock,
    createPswp,
    fakeSrc,
    CreatePswpOptions,
    createProtoPswp,
} from './util'

import { createLocalVue } from '@vue/test-utils'
import VuePswipe from '@/main'
import Vue, { ComponentOptions } from 'vue'
import { PswpDirectiveOptions } from '@/type'

describe('global function', () => {
    beforeEach(() => {
        expect(PhotoSwipe).not.toHaveBeenCalled()
    })

    afterEach(() => {
        PhotoSwipe.mockClear()
    })

    it('global options', () => {
        const bgOpacity = 0.5
        const optionsLocalVue = createLocalVue()
        optionsLocalVue.use(VuePswipe, {
            bgOpacity,
        })

        createPswp({
            localVue: optionsLocalVue,
        })

        const receiveOptions = PhotoSwipeMock.getReceiveOptions()
        expect(receiveOptions.bgOpacity).toBe(bgOpacity)
    })

    it('allow listening to the original PhotoSwipe event', () => {
        const mock = jest.fn()

        createPswp({
            listeners: {
                gettingData: mock,
            },
        })

        expect(mock).toBeCalled()
    })

    describe('history mode', () => {
        const createHashPswp = (hashPath: string, options?: CreatePswpOptions) => {
            window.history.pushState({}, '', hashPath)

            const wrapper = createPswp({
                withClick: false,
                attachToDocument: true,
                propsData: {
                    options: {
                        history: true,
                    },
                },
                ...options,
            })

            document.body.appendChild(wrapper.vm.$el)

            return wrapper
        }

        it('with history hash', (done) => {
            const wrapper = createHashPswp('/#&gid=1&pid=1')

            setTimeout(() => {
                expect(PhotoSwipe).toBeCalled()
                wrapper.destroy()
                done()
            })
        })

        it('with custom hash', (done) => {
            const customPid = 'custom-first-id'

            const wrapper = createHashPswp(`/#&gid=1&pid=${customPid}`, {
                defaultSlots:
                `
                    <img 
                        src="${fakeSrc}" 
                        data-pswp-src="${fakeSrc}"
                        data-pswp-pid="${customPid}"
                    />
                `,
                propsData: {
                    options: {
                        history: true,
                        galleryPIDs: true,
                    },
                },
            })

            setTimeout(() => {
                expect(PhotoSwipe).toBeCalled()
                wrapper.destroy()
                done()
            })
        })
    })

    describe('v-pswp directive', () => {
        const defaultPswpItem = {
            src: 'https://farm4.staticflickr.com/3894/15008518202_c265dfa55f_h.jpg',
            size: '1600x1600',
            msrc: 'https://farm4.staticflickr.com/3894/15008518202_b016d7d289_m.jpg',
            title: 'this is dummy caption',
            pid: 'custom-pid',
        }

        const createDirectivePswp = ({
            pswpItem = defaultPswpItem,
            ...slotOptions
        }: createDirectivePswpArgs = {}) =>
            createPswp({
                defaultSlots: {
                    data: () => ({
                        pswpItem,
                    }),
                    template:
                    `
                        <img
                            :src="pswpItem.msrc"
                            v-pswp="pswpItem"
                        />
                    `,
                    ...slotOptions,
                },
            })


        it('transform to custom dataset', () => {
            const wrapper = createDirectivePswp()

            expect(wrapper.find('img').vm.$el).toMatchSnapshot()
        })

        it('data-pswp-src shorthand', () => {
            const wrapper = createDirectivePswp({
                pswpItem: fakeSrc,
            })

            const el = wrapper.find('img').vm.$el as HTMLElement

            expect(el.dataset.pswpSrc).toBe(fakeSrc)
        })

        it('update directive value', () => {
            const anotherPswpItem = {
                src: 'https://farm6.staticflickr.com/5591/15008867125_b61960af01_h.jpg',
                size: '1600x1068',
                msrc: 'https://farm6.staticflickr.com/5591/15008867125_68a8ed88cc_m.jpg',
                title: 'this is another dummy caption',
                pid: 'another-custom-pid',
            }

            const wrapper = createDirectivePswp({
                methods: {
                    changePswpItem() {
                        this.pswpItem = anotherPswpItem
                    },
                },
            })

            const img = wrapper.find('img')
            const vm = img.vm as ImageComponent

            vm.changePswpItem()
            expect(vm.$el).toMatchSnapshot()
        })
    })

    describe('Vue.prototype.$Pswp', () => {
        it('open()', () => {
            createProtoPswp()
            expect(PhotoSwipe).toBeCalled()
        })
    })
})

interface ImageComponent extends Vue {
    pswpItem: PswpDirectiveOptions
    changePswpItem: () => void
}

type ImageOptions = ComponentOptions<ImageComponent>

interface createDirectivePswpArgs extends ImageOptions {
    pswpItem?: string | PswpDirectiveOptions
}
