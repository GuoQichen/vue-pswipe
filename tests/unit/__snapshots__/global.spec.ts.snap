// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`global function v-pswp directive transform to custom dataset 1`] = `
<img
  data-pswp-msrc="https://farm4.staticflickr.com/3894/15008518202_b016d7d289_m.jpg"
  data-pswp-pid="custom-pid"
  data-pswp-size="1600x1600"
  data-pswp-src="https://farm4.staticflickr.com/3894/15008518202_c265dfa55f_h.jpg"
  data-pswp-title="this is dummy caption"
  src="https://farm4.staticflickr.com/3894/15008518202_b016d7d289_m.jpg"
/>
`;

exports[`global function v-pswp directive update directive value 1`] = `
<img
  data-pswp-msrc="https://farm6.staticflickr.com/5591/15008867125_68a8ed88cc_m.jpg"
  data-pswp-pid="another-custom-pid"
  data-pswp-size="1600x1068"
  data-pswp-src="https://farm6.staticflickr.com/5591/15008867125_b61960af01_h.jpg"
  data-pswp-title="this is another dummy caption"
  src="https://farm6.staticflickr.com/5591/15008867125_68a8ed88cc_m.jpg"
/>
`;
